src/arp.rs

```diff
@@ -1,65 +1,300 @@
-use std::ptr::copy_nonoverlapping;
+use crate::pv::PvPacket;
+use pnet::datalink::MacAddr;
 
-use crate::pv::{PvPacket};
-use pnet::datalink::{MacAddr};
+pub fn endian16(ptr: *const u8) -> u16 {
+    (unsafe { *(ptr.offset(1)) } as u16) << 8 | unsafe { *ptr } as u16
+}
 
-pub fn gen_arp_response_packet(packet: &mut PvPacket, src_mac_address: &MacAddr) {
-    const MAC_ADDRESS_LEN: usize = 6;
-    const ARP_RSPN_PACKET_SIZE: u32 = 42;
+pub fn endian32(ptr: *const u8) -> u32 {
+    unsafe {
+        ((*ptr.offset(3) as u32) << 24)
+            | ((*ptr.offset(2) as u32) << 16)
+            | ((*ptr.offset(1) as u32) << 8)
+            | ((*ptr.offset(0) as u32) << 0)
+    }
+}
 
-    // copy dest MAC
+pub fn endian48(ptr: *const u8) -> u64 {
+    unsafe {
+        ((*ptr.offset(5) as u64) << 40)
+            | ((*ptr.offset(4) as u64) << 32)
+            | ((*ptr.offset(3) as u64) << 24)
+            | ((*ptr.offset(2) as u64) << 16)
+            | ((*ptr.offset(1) as u64) << 8)
+            | ((*ptr.offset(0) as u64) << 0)
+    }
+}
+
+pub fn endian64(ptr: *const u8) -> u64 {
     unsafe {
-        copy_nonoverlapping(packet.buffer.offset((packet.start + 6) as isize), packet.buffer.offset(packet.start as isize), MAC_ADDRESS_LEN);
+        ((*ptr.offset(7) as u64) << 56)
+            | ((*ptr.offset(6) as u64) << 48)
+            | ((*ptr.offset(5) as u64) << 40)
+            | ((*ptr.offset(4) as u64) << 32)
+            | ((*ptr.offset(3) as u64) << 24)
+            | ((*ptr.offset(2) as u64) << 16)
+            | ((*ptr.offset(1) as u64) << 8)
+            | ((*ptr.offset(0) as u64) << 0)
+    }
+}
+
+pub fn macAddr_to_u64(mac: &MacAddr) -> u64 {
+    let mut ret: u64 = 0;
+    let mut n = 0;
+
+    let mac = mac.octets();
+    for i in mac.iter() {
+        ret |= (*i as u64) << (n * 8);
+        n += 1;
+    }
+    ret
+}
+
+fn set_bits(ptr: *mut u8, value: u64, len: u32) {
+    let mut n = 0;
+    unsafe {
+        while len > n {
+            *ptr.offset(n as isize) = (value >> (n * 8)) as u8;
+            n += 1;
+        }
+    }
+}
+
+pub fn check(packet: &PvPacket, pos: u32, size: u32) -> bool {
```

Comment:
I think `Result` type for return value looks better because function name having boolean type for return is usually like this: `is_valid()`, `is_err()`